central/declarativeconfig/manager_impl.go

```diff
@@ -215,8 +246,127 @@ func (m *managerImpl) reconcileTransformedMessages(transformedMessagesByHandler
+		})
+
+		if currentDeclarativeConfigErrors >= threshold {
+			m.declarativeConfigErrorReporter.UpdateIntegrationHealthAsync(&storage.IntegrationHealth{
+				Id:            messageID,
+				Name:          stringutils.FirstNonEmpty(extractNameFromProtoMessage(message), messageID),
+				Type:          storage.IntegrationHealth_DECLARATIVE_CONFIG,
+				Status:        storage.IntegrationHealth_UNHEALTHY,
+				ErrorMessage:  err.Error(),
+				LastTimestamp: timestamp.TimestampNow(),
+			})
+		}
+	} else {
+		var currentDeclarativeConfigErrors int32
+		concurrency.WithRLock(&m.declarativeConfigErrorsLock, func() {
+			currentDeclarativeConfigErrors = m.errorsPerDeclarativeConfig[messageID]
+		})
+
+		if currentDeclarativeConfigErrors > 0 {
+			concurrency.WithLock(&m.declarativeConfigErrorsLock, func() {
+				// Ensure the error count hasn't updated in the mean time.
+				if m.errorsPerDeclarativeConfig[messageID] == currentDeclarativeConfigErrors {
+					m.errorsPerDeclarativeConfig[messageID] = 0
+				}
+			})
+		}
+
+		m.declarativeConfigErrorReporter.UpdateIntegrationHealthAsync(&storage.IntegrationHealth{
+			Id:            messageID,
+			Name:          stringutils.FirstNonEmpty(extractNameFromProtoMessage(message), messageID),
+			Type:          storage.IntegrationHealth_DECLARATIVE_CONFIG,
+			Status:        storage.IntegrationHealth_HEALTHY,
+			ErrorMessage:  "",
+			LastTimestamp: timestamp.TimestampNow(),
+		})
+	}
+
+	log.Debugf("Error within reconciliation for %+v: %v", message, err)
+}
+
+// updateHandlerHealth will update the health status of a handler using the integrationhealth.Reporter.
+func (m *managerImpl) updateHandlerHealth(handlerID string, err error) {
+	var errMsg string
+	if err != nil {
+		errMsg = err.Error()
+	}
+
+	m.declarativeConfigErrorReporter.UpdateIntegrationHealthAsync(&storage.IntegrationHealth{
+		Id:            handlerID,
+		Name:          fmt.Sprintf("Config Map %s", path.Base(handlerID)),
+		Type:          storage.IntegrationHealth_DECLARATIVE_CONFIG,
+		Status:        utils.IfThenElse(err != nil, storage.IntegrationHealth_UNHEALTHY, storage.IntegrationHealth_HEALTHY),
+		ErrorMessage:  errMsg,
+		LastTimestamp: timestamp.TimestampNow(),
+	})
+}
+
+func (m *managerImpl) registerHealthForMessage(messages ...proto.Message) {
+	for _, message := range messages {
+		messageID := extractIDFromProtoMessage(message)
+		messageName := extractNameFromProtoMessage(message)
+
+		if err := m.declarativeConfigErrorReporter.Register(messageID, stringutils.FirstNonEmpty(messageName, messageID),
+			storage.IntegrationHealth_DECLARATIVE_CONFIG); err != nil {
+			log.Errorf("Error registering health status for declarative config %+v: %v", message, err)
+		}
+	}
+}
+
+// Helpers.
+
+func extractIDFromProtoMessage(message proto.Message) string {
+	// Special case, as the group specifies the ID nested within the groups properties.
+	if group, ok := message.(*storage.Group); ok {
```

Comment:
I added the abstraction, although I'm not sure how much actual value it currently adds, but I suppose when we get to the point where we have more proto types it may be come beneficial.